src/parse/function.rs

```diff
@@ -0,0 +1,97 @@
+
+// TODO:
+// - Add documentation
+// - Add more tests
+// - Standardized user facing API.
+
+use super::utils::*;
+use super::*;
+use std::iter::Peekable;
+
+#[derive(Debug)]
+struct Function {
+    visibility: Visibility,
+    is_async: bool,
+    is_unsafe: bool,
+
+    name: String,
+    generics: Option<Generics>,
+
+    // TODO:
+    // args: Vec<FnArg>,
+    // where_cl : WhereClause,
+    // ret_ty : ReturnType,
+    // body: FnBody,
+
+    #[allow(dead_code)]
+    rest: TokenStream, // For debugging purposes
+}
+
+impl Function {
+    pub(crate) fn take(input: &mut Peekable<impl Iterator<Item = TokenTree>>) -> Self {
+        let visibility = Visibility::take(input);
+        let is_async = consume_ident_if(input, "async").is_some();
+        let is_unsafe = consume_ident_if(input, "unsafe").is_some();
+
+        // Ignore everything until `fn` keyword
+        let _ = input.skip_while(|tt| tt.to_string() != "fn").next();
+
+        let name = consume_ident(input)
+            .expect("Expected: `fn <name>`")
```

Comment:
This should probably return an error instead of `.expect()`

Reply:
Maybe, I thought that its fine to panic... 

Becouse [documentation about procedural-macros](https://doc.rust-lang.org/reference/procedural-macros.html) said that,

> they must either return syntax, panic, or loop endlessly, 
> Panics are caught by the compiler and are turned into a compiler error

Reply:
`virtue` is a library that is used by other people to create proc macros. In e.g. [bincode](https://github.com/bincode-org/bincode/blob/trunk/derive/src/lib.rs#L10) we specifically want to catch these errors and give a proper error at the correct location, instead of showing a "panic has occurred during the build".